### PR TITLE
Use a macro to declare required functions

### PR DESCRIPTION
**Pull request by oxinabox**
This seemed an interesting exercise in metaprogramming.
I wanted a list of which things would error if i didn't define them.
Vs which ones had defaults defined.
ergo this PR.

```julia-repl

julia> xx = AbstractFBCModels.required_functions()
[1] reactions(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[2] n_reactions(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[3] metabolites(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[4] n_metabolites(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[5] genes(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[6] n_genes(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[7] stoichiometry(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[8] bounds(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[9] objective(a::AbstractFBCModels.AbstractFBCModel) @ AbstractFBCModels none:0
[10] reaction_stoichiometry(a::AbstractFBCModels.AbstractFBCModel, reaction_id::String) @ AbstractFBCModels none:0
[11] save(a::AbstractFBCModels.AbstractFBCModel, path::String) @ AbstractFBCModels none:0
```

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -3,18 +3,40 @@ import Downloads: download
...
 
 Provide a `methodswith`-style listing of accessors that the model implementors
-should implement.
+may implement.
 
 For typesystem reasons, the list **will not contain** methods for
 [`save`](@ref) and [`filename_extensions`](@ref) that dispatch on type objects.
 You should implement these as well.
+
+See also [`required_functions`](@ref) for the minimal list that must be implemented.
```
Comment (exaexa) [inline src/utils.jl:39]:
```suggestion
See also [`required_accessors`](@ref) for the minimal list that must be implemented.
```

File: src/utils.jl [hunk 2]
```diff
@@ -27,6 +49,30 @@ function accessors()
...
+must implement to have a functional `AbstractFBCModel`.
+constrast this to the longer list of items that are returned by [`accessors`](@ref).
+The extra elements have sensible defaults based on these required functions.
+Where-as not defining these methods will result in errors.
+Though depending on your models capability relying on those defaults may mean some
+functionality is hidden.
+(e.g. default [`coupling`](@ref) if you don't implement that is to assume none)
+
+"""
+function required_functions()
```
Comment (exaexa) [inline src/utils.jl:65]:
```suggestion
function required_accessors()
```

File: src/utils.jl [hunk 3]
```diff
@@ -27,6 +49,30 @@ function accessors()
...
+    ms = Method[]
+    for f = REQUIRED_FUNCTIONS
```
Comment (exaexa) [inline src/utils.jl:67]:
```suggestion
    for f = REQUIRED_ACCESSORS
```

File: src/utils.jl [hunk 4]
```diff
@@ -3,18 +3,40 @@ import Downloads: download
 import SHA: sha256
 import InteractiveUtils: methodswith
 
+
+const REQUIRED_FUNCTIONS = Function[]
```
Comment (exaexa) [inline src/utils.jl:7]:
```suggestion
const REQUIRED_ACCESSORS = Function[]
```

File: src/utils.jl [hunk 5]
```diff
@@ -3,18 +3,40 @@ import Downloads: download
...
+    name isa Symbol || error("malformed signature definition")
+    
+    model_arg_ex = call_ex.args[2]
+    model_arg_ex.head == :(::) || error("malformed signature definition")
+    model_arg = model_arg_ex.args[1]
+    model_arg isa Symbol || error("malformed signature definition")
+
+    return esc(quote
+        Base.@__doc__ $call_ex = $unimplemented(typeof($model_arg), $(Meta.quot(name))) 
+        push!(REQUIRED_FUNCTIONS, $name)
```
Comment (exaexa) [inline src/utils.jl:21]:
```suggestion
        push!(REQUIRED_ACCESSORS, $name)
```

File: src/utils.jl [hunk 6]
```diff
@@ -27,6 +49,30 @@ function accessors()
...
+$(TYPEDSIGNATURES)
+
+Provide a `methodswith`-style listing of functions that the model implementors
+must implement to have a functional `AbstractFBCModel`.
+constrast this to the longer list of items that are returned by [`accessors`](@ref).
+The extra elements have sensible defaults based on these required functions.
+Where-as not defining these methods will result in errors.
+Though depending on your models capability relying on those defaults may mean some
+functionality is hidden.
+(e.g. default [`coupling`](@ref) if you don't implement that is to assume none)
```
Comment (exaexa) [inline src/utils.jl:62]: slight restyle:
```suggestion
Provide a `methodswith`-style listing of functions that the model implementors
must implement to have a functional `AbstractFBCModel`.

The output listing is a subset of the longer list returned by
[`accessors`](@ref). Some of the accessors may have sensible defaults -- e.g.,
it is safe to assume an empty default [`coupling`](@ref) if the functions are
not defined. The accessors which do not possess a natural default and thus
*must be defined* (and trying to use a model without them will almost certainly
cause runtime errors) are exactly the ones listed by this function.
```

**Review by exaexa — COMMENTED** [review]
Looks all good to me, I changed the name of the `_functions` to `_accessors` in a few places to make the code obviously related to accessors, and rewrote the docstring to follow a slightly simpler logical explanation.

If you're okay with the edits we can release this asap.

Anyway we wanted to have something like this for a long time already, so thanks a lot!

File: docs/src/utilities.jl [hunk 7]
```diff
@@ -14,6 +14,8 @@ A.accessors()
 
 @test length(A.accessors()) == 35 #src
 
+@test length(A.required_functions()) == 11
```
Comment (exaexa) [inline docs/src/utilities.jl:17]:
```suggestion
@test length(A.required_accessors()) == 11
```

Comment (exaexa) [reply docs/src/utilities.jl:17]: (the #src is needed to dodge the literate/documenter rendering)

**Comment by oxinabox** [comment]
I went back and forth on the name several times while writing this.
Since it does also include `save` which isn't really an accessor.

**Review by exaexa — APPROVED** [review]
looks all good

**Pull request merged**